proposals/compile-time-dependency-resolution.md

```diff
@@ -0,0 +1,518 @@
+  }
+
+  override fun Group<A>.save() {
+    this.items.map { repoA.run { it.save() } }
+  }
+}
+```
+
+The above extension provides evidence of a `Repository<Group<A>>` as long as there is a `Repository<A>` in scope. Call site would be like:
+
+```kotlin
+fun fetchGroup<A>(with repo: GroupRepository<A>) = repo.loadAll()
+
+fun main() {
+  fetchGroup<User>() // Succeeds! There's evidence of Repository<Group<A>> and Repository<User> provided in scope.
+  fetchGroup<Coin>() // Fails! There's evidence of Repository<Group<A>> but no evidence of `Repository<Coin>` available.
+}
+```
+
+We believe the encoding proposed above fits nicely with Kotlin's philosophy of extensions and will reduce the boilerplate compared to other langs that also support compile time dependency resolution.
+
+## Language changes
+
+* Add `with` to require evidence of extensions in both function and class/object declarations.
+* Add `extension` to provide instance evidence for a given interface.
+
+Usage of these language changes are demonstrated by the previous and below examples:
+
+#### Class constraint
+
+```kotlin
+extension class GroupRepository<A>(with R: Repository<A>) : Repository<Group<A>> {
+  /* ... */
+}
+```
+
+#### Function constraint
+
+```kotlin
+fun <A> fetch(id: String, with R: Repository<A>): A = loadById(id) // function position using parameter "R"
+```
+
+#### Extension evidence using an Object
+
+```kotlin
+extension object UserRepository : Repository<User> {
+
+  val storedUsers: MutableMap<Int, User> = mutableMapOf() // e.g. users stored in a DB
+
+  override fun loadAll(): Map<Int, User> {
+    return storedUsers
+  }
+
+  override fun loadById(id: Int): User? {
+    return storedUsers[id]
+  }
+
+  override fun User.save() {
+    storedUsers[this.id] = this
+  }
+}
+```
+
+#### Extension evidence using a Class
+
+```kotlin
+extension class UserRepository: Repository<User> {
+
+  val storedUsers: MutableMap<Int, User> = mutableMapOf() // e.g. users stored in a DB
+
+  override fun loadAll(): Map<Int, User> {
+    return storedUsers
+  }
+
+  override fun loadById(id: Int): User? {
+    return storedUsers[id]
+  }
+
+  override fun User.save() {
+    storedUsers[this.id] = this
+  }
+}
+```
+
+## Extension resolution order
+
+Classical interfaces only permit their implementation at the site of a type definition. Compiler extension resolution pattern typically relax this rule and **allow extension evidences be declared outside of the type definition**. When relaxing this rule it is important to preserve the coherency we take for granted with classical interfaces.
+
+For those reasons constraint interfaces must be provided in one of the following scopes (in strict resolution order):
+
+1. Arguments of the caller function.
+2. Companion object for the target type (User).
+3. Companion object for the constraint interface we're looking for (Repository).
+4. Subpackages of the package where the target type (User) to resolve is defined.
```

Comment:
Subpackages. So including across modules?

Reply:
Nope. When instances are defined in subpackages of type or type class, they need to be marked as internal.

Reply:
Should we explicitly add that remark here?

Reply:
Yep definitely. Sorry, resolved this one too early. I'll add that now.

Reply:
Fixed in 24b9dfe.